### PR TITLE
166 percentile blend

### PR DESCRIPTION
**Pull request by bayliffe**
Rebased to update TRAVIS environment to ensure tests pass.

Replaces #220 (which replaced #212).

### CODE REVIEW

File: lib/improver/tests/weighted_blend/test_weighted_blend.py [hunk 1]
```diff
@@ -55,7 +57,18 @@ def example_coord_adjust(pnts):
...
+    """Test the repr method."""
+
+    def test_basic(self):
+        """Test that the __repr__ returns the expected string."""
+        result = str(WeightedBlend('time'))
+        msg = '<WeightedBlend: coord = time>'
+        self.assertEqual(result, msg)
+
+
+class TestWeightedBlend(IrisTest):
```
Comment (gavinevans) [inline lib/improver/tests/weighted_blend/test_weighted_blend.py]: I think this should be named `Test_process`.

File: lib/improver/tests/weighted_blend/test_blending_utilities.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for the weighted_blend.BasicWeightedAverage plugin."""
```
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: This file is also named wrongly.

File: lib/improver/tests/weighted_blend/test_blending_utilities.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    """Test the repr method."""
+
+    def test_basic(self):
+        """Test that the __repr__ returns the expected string."""
+        result = str(BlendingUtilities())
+        msg = '<BlendingUtilities>'
+        self.assertEqual(result, msg)
+
+
+class TestBlendingUtilities(IrisTest):
```
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: Yes, I think these need a lot of splitting up.

File: lib/improver/tests/weighted_blend/test_blending_utilities.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        cube = Cube(data, standard_name="precipitation_amount",
+                    units="kg m^-2 s^-1")
+        cube.add_dim_coord(DimCoord(np.linspace(-45.0, 45.0, 2), 'latitude',
+                                    units='degrees'), 1)
+        cube.add_dim_coord(DimCoord(np.linspace(120, 180, 2), 'longitude',
+                                    units='degrees'), 2)
+        time_origin = "hours since 1970-01-01 00:00:00"
+        calendar = "gregorian"
+        tunit = Unit(time_origin, calendar)
+        cube.add_aux_coord(AuxCoord([402192.5, 402193.5],
```
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: I think time should be added as a DimCoord

File: lib/improver/tests/weighted_blend/test_blending_utilities.py (hunk 2 above)
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: Yes

File: lib/improver/tests/weighted_blend/test_weighted_blend.py (hunk 1 above)
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_weighted_blend.py]: I think this is something that is causing conflicts as I changed this to be Test_process during Technical Debt Day

File: lib/improver/weighted_blend.py [hunk 5]
```diff
@@ -33,13 +33,208 @@
...
+                    result = np.moveaxis(result, 0, perc_dim)
+        return result
+
+    @staticmethod
+    def blend_percentiles(perc_values, percentiles, weights):
+        """ Blend percentiles function
+
+        Args:
+            perc_values : np.array
+                   Array containing the percentile values to blend
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Changed to this: 

`Array containing the percentile values to blend, with
shape (length of coord to blend, num of percentiles)`

File: lib/improver/tests/weighted_blend/test_blending_utilities.py (hunk 2 above)
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: Fixed

File: lib/improver/tests/weighted_blend/test_blending_utilities.py (hunk 3 above)
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: Fixed with basic docstrings, which will need refining when we review the unit tests

File: lib/improver/weighted_blend.py [hunk 6]
```diff
@@ -33,13 +33,208 @@
...
+        return result
+
+    @staticmethod
+    def basic_weighted_average(cube, weights, coord, coord_dim):
+        """Calculate weighted mean across the chosen coord
+
+        Args:
+            cube : iris.cube.Cube
+                   Cube to blend across the coord.
+            weights: Optional list or np.array of weights
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Changed

File: lib/improver/weighted_blend.py [hunk 7]
```diff
@@ -33,13 +33,208 @@
...
+        return result
+
+    @staticmethod
+    def blend_percentile_cube(cube, weights, coord, coord_dim, perc_coord):
+        """ Blend together percentile cube
+
+        Args:
+            cube : iris.cube.Cube
+                   Cube to blend across the coord.
+            weights: Optional list or np.array of weights
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Changed

File: lib/improver/weighted_blend.py [hunk 8]
```diff
@@ -33,13 +33,208 @@
...
+    @staticmethod
+    def blend_percentile_aggregate(data, axis,
+                                   arr_percent, arr_weights, perc_dim):
+        """ Blend percentile aggregate function
+
+        Args:
+            data : np.array
+                   Array containing the data to blend
+            axis : integer
+                   The index of the coordinate dimension in the cube.
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Added a sentence to explain this.

File: lib/improver/weighted_blend.py [hunk 9]
```diff
@@ -33,13 +33,208 @@
...
+            arr_percent: np.array
+                     Array of percentile values e.g
+                     [0, 20.0, 50.0, 70.0, 100.0],
+                     same size as the percentile dimension of data.
+            arr_weights: np.array
+                     Array of weights, same size as the axis dimension of data.
+            perc_dim : integer
+                     The index of the perecentile coordinate
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Fixed

File: lib/improver/weighted_blend.py [hunk 10]
```diff
@@ -33,13 +33,208 @@
...
+            cube : iris.cube.Cube
+                   Cube to blend across the coord.
+            weights: Optional list or np.array of weights
+                     or None (equivalent to equal weights)
+            coord : string
+                     The name of a coordinate dimension in the cube.
+            coord_dim : tuple
+                     The index of the coordinate dimension in the cube.
+            perc_coord : iris.cube.DimCoord
+                     The perecentile coordinate
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Fixed

File: lib/improver/weighted_blend.py [hunk 11]
```diff
@@ -33,13 +33,208 @@
...
+            axis += data.ndim
+        # Firstly ensure axis coordinate and percentile coordinate
+        # are indexed as the first and second values in the data array
+
+        data = np.moveaxis(data, [perc_dim, axis], [1, 0])
+
+        # Determine the rest of the shape
+        shape = data.shape[2:]
+        result = None
+        if shape:
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: This doesn't work, and I don't think this is the correct behaviour. I'll try and fix it.

File: lib/improver/weighted_blend.py [hunk 12]
```diff
@@ -33,13 +33,208 @@
...
+                                arr_percent=percentiles,
+                                arr_weights=weights,
+                                perc_dim=perc_dim)
+        return result
+
+    @staticmethod
+    def blend_percentile_aggregate(data, axis,
+                                   arr_percent, arr_weights, perc_dim):
+        """ Blend percentile aggregate function
+
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Added a bit more detail, and a bigger docstring at the top of the aggregator class

File: lib/improver/weighted_blend.py [hunk 13]
```diff
@@ -33,13 +33,208 @@
...
+                    result = np.moveaxis(result, 0, perc_dim-1)
+            else:
+                if perc_dim != 0:
+                    result = np.moveaxis(result, 0, perc_dim)
+        return result
+
+    @staticmethod
+    def blend_percentiles(perc_values, percentiles, weights):
+        """ Blend percentiles function
+
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Done, along with top level description

File: lib/improver/weighted_blend.py (hunk 11 above)
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: I have removed the if shape statement, and added tests for 2D cubes

File: lib/improver/weighted_blend.py [hunk 14]
```diff
@@ -33,13 +33,208 @@
...
+            result = np.zeros(input_shape[1:])
+            # Loop over the flatten data
+            for i in range(data.shape[-1]):
+                result[:, i] = (
+                    BlendingUtilities.blend_percentiles(data[:, :, i],
+                                                        arr_percent,
+                                                        arr_weights))
+        # Reshape the data and put the percentile dimension
+        # back in the right place
+        if arr_percent.shape > (1,):
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Seeing as the blending will fail if only one percentile is present in the cube, and I've put in an exception to catch this case, I've removed this if statement from the code.

File: lib/improver/weighted_blend.py [hunk 15]
```diff
@@ -91,8 +298,27 @@ def process(self, cube, weights=None):
...
-        weights_array = None
+
+        # Check to see if the data is percentile data
+        perc_coord = None
+        perc_dim = None
+        perc_found = 0
+        for coord in cube.coords():
+            if coord.name().find('percentile') >= 0:
+                perc_found += 1
+                perc_coord = coord
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: I think Caroline's way is fine. We could pass in the percentile coordinate, but I think this way works fine.

File: lib/improver/weighted_blend.py [hunk 16]
```diff
@@ -102,16 +294,52 @@ def process(self, cube, weights=None):
...
+                                    arr_weights=weights,
+                                    perc_dim=perc_dim)
+
+        # Else do a simple weighted average
+        else:
+            # Equal weights are used as default.
+            weights_array = None
+            # Else broadcast the weights to be used by the aggregator.
+            if weights is not None:
+                weights_array = iris.util.broadcast_to_shape(np.array(weights),
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: I'm slightly concerned we are not treating the case where a list (rather than an array) is passed in as weights.  There are some unit tests to test what happens if a list is passed in, but I think we are doing something different if we are blending probabilities compared to percentiles. I think we should probably not be recasting the input as an array. What do you think @carolinejjones ?

File: lib/improver/weighted_blend.py [hunk 17]
```diff
@@ -33,13 +33,208 @@
 
 import numpy as np
 import iris
+from iris.analysis import Aggregator
 
 
-class BasicWeightedAverage(object):
-    """Apply a Basic Weighted Average to a cube."""
+class BlendingUtilities(object):
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: I've left this as a smaller class inside this file

File: lib/improver/weighted_blend.py [hunk 18]
```diff
@@ -33,13 +33,208 @@
...
+                     The name of a coordinate dimension in the cube.
+            coord_dim : tuple
+                     The index of the coordinate dimension in the cube.
+
+        Returns:
+            result : iris.cube.Cube
+                     containing the weighted mean across the chosen coord
+        """
+        # Supply weights as an array of weights whose shape matches the cube.
+        weights_array = None
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: I think this is ok, although see my other comment on the treatment of weights input as weights

File: lib/improver/weighted_blend.py [hunk 19]
```diff
@@ -33,13 +33,208 @@
...
+            perc_coord : iris.cube.DimCoord
+                     The perecentile coordinate
+
+        Returns:
+            result : iris.cube.Cube
+                     containing the weighted percentile blend
+                     across the chosen coord
+        """
+
+        percentiles = np.array(perc_coord.points, dtype=float)
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: I think I've got rid of this line, but it might be needed to make sure the percentile values are floats

File: lib/improver/weighted_blend.py [hunk 20]
```diff
@@ -102,12 +328,20 @@ def process(self, cube, weights=None):
...
                 raise ValueError(msg)
-            weights_array = iris.util.broadcast_to_shape(np.array(weights),
-                                                         cube.shape,
-                                                         collapse_dim)
-        # Calculate the weighted average.
-        result = cube.collapsed(self.coord,
-                                iris.analysis.MEAN, weights=weights_array)
+
+        # Blend the cube across the coordinate
+        if perc_coord is not None:
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: I'm happy to leave it working as Caroline intended

File: lib/improver/tests/weighted_blend/test_blending_utilities.py (hunk 4 above)
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: I've changed this

**Comment by fionaRust** [comment]
I've had a look at the data from the new CLI test. It's hard to know what the expected output should be but it looks like it could be reasonable. Link to ipython notebook on Jira

File: lib/improver/weighted_blend.py [hunk 21]
```diff
@@ -57,20 +218,35 @@ def __init__(self, coord, coord_adjust=None):
...
         Returns:
             result : iris.cube.Cube
-                     containing the weighted mean across the chosen coord
+                     containing the weighted blend across the chosen coord.
+
+        Raises:
+            ValueError 1 : first argument not a cube.
+            ValueError 2 : The coord is not a coord with the cube.
+            ValueError 3 : If there is a percentile coord it is not a
+                           dimension coord in the cube.
```
Comment (carolinejjones) [inline lib/improver/weighted_blend.py]: If we are leaving this section in we need to add the ValueError when the percentile coord has less than two points

File: lib/improver/weighted_blend.py [hunk 22]
```diff
@@ -57,20 +218,35 @@ def __init__(self, coord, coord_adjust=None):
...
+            ValueError 4 : If there are more than one percentile coords
+                           in the cube.
+            ValueError 5 : weights shape do not match the dimension
+                           of the coord.
+            Warning : coord not a dimension within the cube.
```
Comment (carolinejjones) [inline lib/improver/weighted_blend.py]: I think looking at the warning message below that this would be meaningful if it reflected the warning message.
i.e.
Warning : Trying to blend across a scalar coordinate with only one value

File: lib/improver/weighted_blend.py (hunk 16 above)
Comment (carolinejjones) [inline lib/improver/weighted_blend.py]: I think we need to recast the input as an array to get this to work properly. I might be wrong. It isn't doing any harm.

File: lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+        weights = np.array([0.8, 0.2])
+        percentiles = np.array([0, 50, 100])
+        perc_data = np.array([[1.0, 2.0], [5.0, 5.0], [10.0, 9.0]])
+        result = PercentileBlendingAggregator.aggregate(
+            perc_data, 1,
+            percentiles,
+            weights, 0)
+        expected_result = np.array([1.0, 5.0, 10.0])
+        self.assertArrayAlmostEqual(result, expected_result)
+
```
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py]: It was mostly to do with checking the code worked with different numbers of dimensions. I removed a check that you had to make sure that the input cube had at least 3 dimensions, but so I wanted to check it did actually work with two input dimensions. But you're right it's not really checking any kind of sensible blending.

File: lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py [hunk 24]
```diff
@@ -0,0 +1,270 @@
...
+        """Test that when two percentiles are provided, the extreme values in
+           the set of thresholds we are blending are returned"""
+        weights = np.array([0.5, 0.5])
+        percentiles = np.array([30., 60.])
+        percentile_values = np.array([[5.0, 8.0], [6.0, 7.0]])
+        result = PercentileBlendingAggregator.blend_percentiles(
+                    percentile_values, percentiles, weights)
+        expected_result = np.array([5.0, 8.0])
+        self.assertArrayAlmostEqual(result, expected_result)
+
```
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py]: Yes, which is why I wanted a test to document that this is what will happen in the case that you don't have any percentiles at the ends of your pdf.  I think it is important to document in the unit tests what happens in edge cases. I think this is the only sensible thing it could do in this case. Whatever the values of the highest and lowest values in your cube, they will will be assigned the maximum and minimum values you had in the data you are blending, so I wanted to write a test to show that is what happens.

File: lib/improver/weighted_blend.py [hunk 25]
```diff
@@ -57,20 +252,31 @@ def __init__(self, coord, coord_adjust=None):
...
-                     or None (equivalent to equal weights)
+                     or None (equivalent to equal weights).
 
         Returns:
             result : iris.cube.Cube
-                     containing the weighted mean across the chosen coord
+                     containing the weighted blend across the chosen coord.
+
+        Raises:
+            ValueError 1 : first argument not a cube.
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: I'm happy to get rid of this section. Although it is in the google style guide: https://google.github.io/styleguide/pyguide.html#Comments

File: lib/improver/tests/weighted_blend/test_WeightedBlend.py [hunk 26]
```diff
@@ -168,8 +219,61 @@ def test_weights_equal_list(self):
...
+        result = plugin.process(perc_cube, weights)
+        expected_result_array = np.reshape(BLENDED_PERCENTILE_DATA1,
+                                           (6, 2, 2))
+        self.assertArrayAlmostEqual(result.data, expected_result_array)
+
+    def test_percentiles_weights_equal_list(self):
+        """Test it works for percentiles with weights [0.8, 0.2]."""
+        coord = "time"
+        plugin = WeightedBlend(coord)
         weights = np.array([0.8, 0.2])
```
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_WeightedBlend.py]: Fixed

File: lib/improver/weighted_blend.py (hunk 25 above)
Comment (carolinejjones) [inline lib/improver/weighted_blend.py]: If it is in the google style guide it would be helpful if there was an example in the wiki pages on good code practice so we knew how to implement this. I agree that labelling them ValueError 1 etc it is not very help and in the long run is likely to make the code hard to maintain.

File: lib/improver/tests/weighted_blend/test_WeightedBlend.py [hunk 27]
```diff
@@ -168,8 +219,61 @@ def test_weights_equal_list(self):
...
+        self.assertArrayAlmostEqual(result.data, expected_result_array)
+
+    def test_basic_weighted_average(self):
+        """Test basic_weighted_average works."""
+        coord = "time"
+        weights = [0.8, 0.2]
+        plugin = WeightedBlend(coord)
         result = plugin.process(self.cube, weights)
         expected_result_array = np.ones((2, 2))*1.2
         self.assertArrayAlmostEqual(result.data, expected_result_array)
```
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_WeightedBlend.py]: Oops, this is one I moved from the other test file. I'll delete it

File: lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py (hunk 24 above)
Comment (carolinejjones) [inline lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py]: Unfortunately this is not a good result. I think it shows that we need at least a warning message that the percentile dimension is not covering the range of the pdf.

File: lib/improver/weighted_blend.py (hunk 16 above)
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Oh yes. Sorry I forgot what this did. I think you're right and there isn't really a problem.

File: lib/improver/tests/weighted_blend/test_WeightedBlend.py [hunk 28]
```diff
@@ -168,8 +219,61 @@ def test_weights_equal_list(self):
...
+        """Test blending percentile cube works with unequal weights."""
+        coord = "time"
+        weights = np.array([0.8, 0.2])
+        perc_cube = percentile_cube()
+        plugin = WeightedBlend(coord)
+        result = plugin.process(perc_cube, weights)
+        expected_result_array = np.reshape(BLENDED_PERCENTILE_DATA2,
+                                           (6, 2, 2))
+        self.assertArrayAlmostEqual(result.data, expected_result_array)
+
```
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_WeightedBlend.py]: These were the two I copied from the other test file, to match the unit test conventions. I guess the only extra things they are testing is the passing in weights as a list rather than an array and passing in weights=None. Seeing as these things are tested in other tests above, I would be happy to remove them.

File: lib/improver/weighted_blend.py (hunk 25 above)
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Yes I think that's true. @benfitzpatrick  could you give some guidance?

I think value error 2 could be removed as iris will raise a sensible error on its own. I think this could be true of value error 1 too. This would mean there are slightly less errors we are raising.

Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Depending on what we decide I'll tidy up this section appropriately.

File: lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py (hunk 24 above)
Comment (fionaRust) [inline lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py]: Ok, I could add that in. At what point would that be raised. Would it only be if we didn't have the 0th and 100th percentiles in the percentile coordinate or would the cut off be before that?

File: lib/improver/weighted_blend.py (hunk 22 above)
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Fixed

File: lib/improver/weighted_blend.py (hunk 21 above)
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Added

File: lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py (hunk 24 above)
Comment (carolinejjones) [inline lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py]: I shall raise an technical debt issue for this as I think we need to look into what to do if there is insufficient data to produce a sensible answer.

Comment (gavinevans) [inline lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py]: @carolinejjones, raising a technical debt issue would be great. It would be useful if you could include a link to  
[ensemble_copula_coupling.py](https://github.com/metoppv/improver/blob/master/lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py), as I'm not sure there's overlap between this problem, and the problems that have been encountered in ECC, regarding when values are required outside the range provided by the input percentiles.

Comment (carolinejjones) [inline lib/improver/tests/weighted_blend/test_PercentileBlendingAggregator.py]: IMPRO-275 raised to investigate this issue.

File: lib/improver/weighted_blend.py [hunk 29]
```diff
@@ -33,13 +33,174 @@
...
+from iris.analysis import Aggregator
 
 
-class BasicWeightedAverage(object):
-    """Apply a Basic Weighted Average to a cube."""
+class PercentileBlendingAggregator(object):
+    """Class for the percentile blending aggregator
+
+       This class implements the method described by Combining Probabilities
+       by Caroline Jones, 2017. This method implements blending in probability
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Done for now, but we still need a better way of doing this.

File: lib/improver/weighted_blend.py [hunk 30]
```diff
@@ -33,13 +33,174 @@
...
+        # Iris aggregators support indexing from the end of the array.
+        if axis < 0:
+            axis += data.ndim
+        # Firstly ensure axis coordinate and percentile coordinate
+        # are indexed as the first and second values in the data array
+        data = np.moveaxis(data, [perc_dim, axis], [1, 0])
+
+        # Determine the rest of the shape
+        shape = data.shape[2:]
+        result = None
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Oops, got rid of an if statement and left it there. Fixed now.

File: lib/improver/weighted_blend.py [hunk 31]
```diff
@@ -33,13 +33,174 @@
...
+        input_shape = [data.shape[0],
+                       data.shape[1],
+                       np.prod(shape, dtype=int)]
+        # Flatten the data that is not percentile or coord data
+        data = data.reshape(input_shape)
+        # Create the resulting data array, which is the shape of the original
+        # data without dimension we are collapsing over
+        result = np.zeros(input_shape[1:])
+        # Loop over the flatten data, i.e. acrosss all the data points in each
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Fixed

File: lib/improver/weighted_blend.py [hunk 32]
```diff
@@ -33,13 +33,174 @@
...
+            if perc_dim != 0:
+                result = np.moveaxis(result, 0, perc_dim)
+        return result
+
+    @staticmethod
+    def blend_percentiles(perc_values, percentiles, weights):
+        """ Blend percentiles function, to calculate the weighted blend across
+            a given axis of percentile data for a single grid point.
+
+        Args:
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Fixed

File: lib/improver/weighted_blend.py [hunk 33]
```diff
@@ -33,13 +33,174 @@
...
+        Returns:
+            result : np.array
+                     containing the weighted percentile blend data
+                     across the chosen coord
+        """
+        # Find the size of the dimension we want to blend over.
+        num = perc_values.shape[0]
+        # Create an array to store the weighted blending pdf
+        combined_pdf = np.zeros((num, len(percentiles)))
+        # Loop over the axis we are blending over finding the values for the
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Fixed

File: lib/improver/weighted_blend.py [hunk 34]
```diff
@@ -57,42 +218,72 @@ def __init__(self, coord, coord_adjust=None):
...
+            ValueError : If there is a percentile coord and it is not a
+                           dimension coord in the cube.
+            ValueError : If there is a percentile dimension with only one
+                            point, we need at least two points in order to do
+                            the blending.
+            ValueError : If there are more than one percentile coords
+                           in the cube.
+            ValueError : If the weights shape do not match the dimension
+                           of the coord we are blending over.
+            Warning : If trying to blend across a scalar coordinate with only
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: Added

File: lib/improver/weighted_blend.py [hunk 35]
```diff
@@ -33,13 +33,174 @@
...
+        for i in range(data.shape[-1]):
+            result[:, i] = (
+                PercentileBlendingAggregator.blend_percentiles(
+                    data[:, :, i], arr_percent, arr_weights))
+        # Reshape the data and put the percentile dimension
+        # back in the right place
+        shape = arr_percent.shape + shape
+        result = result.reshape(shape)
+        if axis < perc_dim:
+            if perc_dim != 1:
```
Comment (fionaRust) [inline lib/improver/weighted_blend.py]: No I think this is right.

Comment (fionaRust) [inline lib/improver/weighted_blend.py]: We've put some more comments in to explain this
